### ISSUE TITLE
A ggplot question: annotating certain points on the plot with their exact (x, y) values

### ISSUE THREAD

**Issue by kaushalmodi** [Vindaar/ggplotnim#37]
Hello,

I am printing an FFT plot in which it is important to know the exact x and y values for the 5 highest x values.

Here's one plot created using ggplotnim:

![image](https://user-images.githubusercontent.com/3578197/74634906-e2974300-5132-11ea-8257-6c4ab61204f9.png)

As you see in the plot, I would like to know the exact X and Y values for that peak around -30MHz (x axis value).

- Is it possible to overlay the (x,y) values for top 5 x values?
- If not that, then may be in a legend kind of thing on the side that lists the 5 highest X values and the associated Y values?

Thanks!

---

PS: btw I am viewing that png in Emacs.. I am wondering if I can make the entire canvas background white so that the axes markers and labels in black don't blend with my default background (grey) in Emacs.

**Comment by Vindaar** [comment]
At the moment there's no exposed API to put text on the plot. However, the underlying code is perfectly capable of that.
I simply need to add `geom_text` (and maybe `geom_label`; it's the same thing but with a colored background for easier reading) and an `annotate` proc.

See here for `geom_text` in `ggplot2`:
https://ggplot2.tidyverse.org/reference/geom_text.html

and `annotate`:
https://ggplot2.tidyverse.org/reference/annotate.html

That means using `annotate` you'd have to calculate the max positions manually. Something like
```nim
let maxVals = df.arrange("x", SortOrder.Descending).take(5)
```
and then make some nice string representation of it (or loop over the 5 each) and put it into `annotate`. That will allow you to place the text wherever you want.

Using `geom_text` you're essentially adding a new geom layer, which will simply draw by printing text. To only get the max 5 values, you'd still have to do some calculation though, e.g. when plotting
```nim
ggplot(df, aes("x", "y")) + 
  geom_line() +
  geom_text(data = maxVals, nudge_y = 0.5) + # the maxVals from above
  ggsave("foo.pdf")
```
because by default the `geom_text` will inherit the data from `ggplot` and thus draw text for all values. We nudge the values a little bit so they don't overlap the line. 
However, if the data points are dense the text probably will still overlap.

I'm not sure which way is more convenient. Personally in plots where I wanted to annotate stuff I was never a big fan of automatically assigning text to data points, because while it'll look nice in one plot in another it'll be a crowded mess again.

We can always add more convenience options for such a use case of course. If you have ideas, feel free to share.

Regarding the background: that's a trivial addition. Thanks for the note. Since I usually save as PDFs I never encountered that problem. The conversion from PDF to PNG (I assume) in emacs automatically adds a white background.


So I would first fix the background issue by adding a `Theme` field for the canvas background and use a white default for PNGs. Then add `annotate`, because that's mostly just exposing the existing functionality. `geom_text` will be last.

**Comment by kaushalmodi** [comment]
> I'm not sure which way is more convenient. Personally in plots where I wanted to annotate stuff I was never a big fan of automatically assigning text to data points, because while it'll look nice in one plot in another it'll be a crowded mess again.

Yes, I can imagine that. The best idea I have is to put tiny circle markers at the points of interest on that line graph (max 5 x values in my example) and annotate them A -> E, and then in a legend show the (x, y) values for each of those points. _I don't even know if that is doable_.

**Comment by Vindaar** [comment]
> _I don't even know if that is doable_.

Heh, since this is all written from scratch it is certainly _doable_. The question is how easy is it and how does it generalize?

One could certainly think of something like a `highlight` functionality, which provides an auto annotate option. Funnily just yesterday Hadley Wickham retweeted a new package for `ggplot2` to highlight stuff: https://twitter.com/yutannihilat_en/status/1229425657160953857

So this kinda got me thinking in that direction anyways, haha. I haven't actually checked out that package though. 

What I imagine is something along the lines of:
- highlight:
  - select subset of data
  - decide on highlighting style (color of points, ellipse/circle around the data etc.)
  - optionally annotate each point by some user defined schema (1, 2, 3; A, B, C; etc)
  - optionally add a legend giving the data for these points

In regards to implementation:
- selection of data is done by user. 
- we can then know if we're drawing a point in the highlighted section and thus change its style or e.g. circle it. Determining location for ellipse / circle is reasonably easy. This will of course result in really ugly plots if some weird data selection is made and the wrong highlighting option is selected. Performing sanity checks on data selection <-> highlighting scheme in general is probably quite hard (but not really important, imo).
- annotating points by schema is just application of `geom_text` behind the curtains. Needs to be implemented first of course.
- Custom legends are easy.

Putting all of this in without making an even bigger mess (*cough*) of the code base is probably a little more of a challenge. Although maybe this can be done in a second pass after the actual plot is created. 


**To summarize**: I think it's not that hard by itself. However, there's probably use cases where this will break, because the user tries to use it in an unforeseen way. Implementing `annotate`, `geom_text` and custom legends get us 90 % of the way with slightly more manual work on the user's part. So that will come first.

**Comment by Vindaar** [comment]
With #41 merged you can now set the canvas background color to whatever you want.

For white you can just use the `theme_opaque` function.

**Comment by Vindaar** [comment]
Once #42 is merged custom annotations can be placed on the plot.

There's some issues in the implementation and it's not amazingly pretty yet, but better to have it now in not so perfect shape than wait an eternity since I fix everything, heh.

See https://github.com/Vindaar/ggplotnim/blob/addCustomAnnotations/recipes.org#plot-with-custom-annotation for an example.

**Comment by kaushalmodi** [comment]
"beauty is in the eye of the beholder"

Those plots are awesome!! I cannot wait to try them out! I'll let you know as soon as I do it.

Thanks.

**Comment by kaushalmodi** [comment]
Can the A, B, C, D, E points put as some kind of markers on that line plot in your example?

**Comment by Vindaar** [comment]
With both white backgrounds and custom annotations now merged and specific highlighting delegated to its own issue, I'll close this one.

**Comment by kaushalmodi** [comment]
Thank you!

The annotation works wonderfully!

This is what I used:

```nim
  # Create a dataframe from the data.
  let
    df = seqsToDf({plotXLabel: x, plotYLabel: y })

  var
    annot: seq[string]

  # https://github.com/Vindaar/ggplotnim/blob/master/recipes.org#plot-with-custom-annotation
  # Create an annotation that prints the largest 5 values of Y axis values
  const
    maxN = 5
  var
    dfFilt = df
               .arrange(plotYLabel, SortOrder.Descending)
  for i in 1 .. maxN:
    const
      ignoreSurrBins = 10
    let
      freq = dfFilt[plotXLabel].vToSeq[0].pretty().parseFloat()
      pwr = dfFilt[plotYLabel].vToSeq[0].pretty().parseFloat()
    if i == 1:
      annot.add &"Fund freq: {freq:8.2f} MHz => {pwr:7.2f} dB"
    else:
      annot.add &"Freq {i}   : {freq:8.2f} MHz => {pwr:7.2f} dB"
    if i < maxN:
      # Filter out ignoreSurrBins bins on each side of the last freq.
      dfFilt = dfFilt.filter(f{plotXLabel < freq-(resMHz*ignoreSurrBins) or
                               plotXLabel > freq+(resMHz*ignoreSurrBins)})
      if dfFilt.len == 0:
        break

  echo &"Saving plot to {plotFilePath} .."
  ggplot(df, aes(plotXLabel, plotYLabel)) +
    theme_opaque() + # https://github.com/Vindaar/ggplotnim/pull/41/files#diff-0c88c9acc6458974b569d4f97da31c85R539
    geom_line() +
    annotate(annot.join("\n"),
             left = 0.58,
             bottom = 0.17,
             # Create a font to use using the `ggplotnim.font` helper.
             font = font(12.0, family = "monospace")) +
    ggtitle($plotCfg.title) +
    ggsave(plotFilePath, width = plotWidth.float, height = plotHeight.float)
```

![image](https://user-images.githubusercontent.com/3578197/75002920-0ceb4800-5434-11ea-86e1-2c2fbc1b4243.png)
